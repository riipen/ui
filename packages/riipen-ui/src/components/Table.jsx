import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";
import withForwardedRef from "../utils/withForwardedRef";

class Table extends React.Component {
  static propTypes = {
    /**
     * What background to render for the table.
     * Defaults to grey100.
     */
    backgroundColor: PropTypes.oneOf([
      "transparent",
      "grey50",
      "grey100",
      "grey200",
      "grey300",
      "grey400"
    ]),

    /**
     * Children to render in the table element.
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element.
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * A reference to the table DOM element.
     */
    forwardedRef: PropTypes.element,

    /**
     * Whether to give all columns in the table equal width.
     */
    layout: PropTypes.oneOf(["auto", "fixed"])
  };

  static defaultProps = {
    backgroundColor: "grey100",
    layout: "auto"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      layout,
      backgroundColor,
      forwardedRef
    } = this.props;

    const theme = this.context;

    const tableBackground = {
      transparent: "transparent",
      grey50: theme.palette.grey[50],
      grey100: theme.palette.grey[100],
      grey200: theme.palette.grey[200],
      grey300: theme.palette.grey[300],
      grey400: theme.palette.grey[400]
    };

    return (
      <React.Fragment>
        <table ref={forwardedRef} className={clsx(classes)} border="0">
          {children}
        </table>
        <style jsx>{`
          table {
            background-color: ${tableBackground[backgroundColor]};
            border-radius: 4px;
            border-collapse: collapse;
            table-layout: ${layout};
            width: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withForwardedRef(withClasses(Table));
