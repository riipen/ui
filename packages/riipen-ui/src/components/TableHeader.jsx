import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class TableHeader extends React.Component {
  static propTypes = {
    /**
     * Children to render in the thead element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const { children, classes, ...other } = this.props;
    return (
      <thead className={clsx(classes)} {...other}>
        {children}
      </thead>
    );
  }
}

export default withClasses(TableHeader);
