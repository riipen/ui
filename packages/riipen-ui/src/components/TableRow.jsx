import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class TableRow extends React.Component {
  static propTypes = {
    /**
     * Children to render in the tr element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * Whether to highlight the row on hover
     */
    hover: PropTypes.bool,

    /**
     * Whether to render the border for the row
     */
    border: PropTypes.bool
  };

  static defaultProps = {
    border: true
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, hover, border } = this.props;

    const theme = this.context;

    return (
      <React.Fragment>
        <tr className={clsx(classes)}>{children}</tr>
        <style jsx>{`
          tr {
            border-bottom: 1px solid
              ${border &&
                (theme.palette.divider
                  ? theme.palette.divider
                  : `transparent`)};
            border-left: 3px solid transparent;
          }

          tr:hover {
            ${hover ? `background-color: ${theme.palette.common.white};` : ""}
            ${hover
              ? `border-left: 3px solid ${theme.palette.secondary.main};`
              : ""}
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(TableRow);
