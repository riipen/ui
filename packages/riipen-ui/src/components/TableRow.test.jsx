import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import defaultTheme from "../themes/default";

import TableRow from "./TableRow";
import Table from "./Table";
import TableHeader from "./TableHeader";

describe("<TableRow>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(
        <Table>
          <TableHeader>
            <TableRow />
          </TableHeader>
        </Table>
      );
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(
      <Table>
        <TableHeader>
          <TableRow />
        </TableHeader>
      </Table>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new TableRow().type.defaultProps;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow />
          </TableHeader>
        </Table>
      );

      expect(wrapper.find("TableRow").props().border).toEqual(
        defaultProps.border
      );
      expect(wrapper.find("TableRow").props().forceHover).toEqual(
        defaultProps.forceHover
      );
    });
  });

  describe("border prop", () => {
    it("sets border with custom color when border is true", () => {
      const border = true;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow border={border} />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .find("JSXStyle")
          .props().dynamic
      ).toContain(defaultTheme.palette.divider);
    });

    it("does not set border when border is false", () => {
      const border = false;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow border={border} />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .find("JSXStyle")
          .props().dynamic
      ).not.toContain(defaultTheme.palette.divider);
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <td>Test</td>;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow>{child}</TableRow>
          </TableHeader>
        </Table>
      );

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the tbody node", () => {
      const classes = ["classOne"];

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow classes={classes} />
          </TableHeader>
        </Table>
      );

      expect(wrapper.find("TableRow").props().classes).toContain(classes[0]);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-tablerow"];

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("forceHover prop", () => {
    it("sets correct class name when forceHover is true", () => {
      const forceHover = true;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow forceHover={forceHover} />
          </TableHeader>
        </Table>
      );

      expect(wrapper.find("tr").hasClass("forceHover")).toEqual(true);
    });
  });

  describe("hover prop", () => {
    it("sets background color when hover is true", () => {
      const hover = true;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow hover={hover} />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .find("JSXStyle")
          .props().dynamic
      ).toContain(`background-color: ${defaultTheme.palette.common.white};`);
    });

    it("does not set background color when hover is false", () => {
      const hover = false;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow hover={hover} />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .find("JSXStyle")
          .props().dynamic
      ).not.toContain(
        `background-color: ${defaultTheme.palette.common.white};`
      );
    });

    it("sets border-left as solid when hover is true", () => {
      const hover = true;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow hover={hover} />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .find("JSXStyle")
          .props().dynamic
      ).toContain(
        `border-left: 3px solid ${defaultTheme.palette.secondary.main};`
      );
    });

    it("does not set border-left as solid when hover is false", () => {
      const hover = false;

      const wrapper = mount(
        <Table>
          <TableHeader>
            <TableRow hover={hover} />
          </TableHeader>
        </Table>
      );

      expect(
        wrapper
          .find("TableRow")
          .find("JSXStyle")
          .props().dynamic
      ).not.toContain(
        `border-left: 3px solid ${defaultTheme.palette.secondary.main};`
      );
    });
  });
});
