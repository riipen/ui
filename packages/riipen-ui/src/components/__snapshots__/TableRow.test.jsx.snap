// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableRow> renders correct snapshot 1`] = `
<ForwardRef(withForwardedRef(withClasses(Table)))>
  <withClasses(Table)
    forwardedRef={null}
  >
    <Table
      backgroundColor="grey50"
      classes={
        Array [
          "riipen",
          "riipen-table",
        ]
      }
      forwardedRef={null}
      layout="auto"
    >
      <table
        border="0"
        className="jsx-1059775881 riipen riipen-table"
      >
        <withClasses(TableHeader)>
          <TableHeader
            classes={
              Array [
                "riipen",
                "riipen-tableheader",
              ]
            }
          >
            <thead
              className="riipen riipen-tableheader"
            >
              <withClasses(TableRow)>
                <TableRow
                  border={true}
                  classes={
                    Array [
                      "riipen",
                      "riipen-tablerow",
                    ]
                  }
                  forceHover={false}
                >
                  <tr
                    className="jsx-4111142428 riipen riipen-tablerow"
                  />
                  <JSXStyle
                    dynamic={
                      Array [
                        "#ddd",
                        "#fff",
                        "#f58632",
                        "",
                        "",
                      ]
                    }
                    id="2808869914"
                  />
                </TableRow>
              </withClasses(TableRow)>
            </thead>
          </TableHeader>
        </withClasses(TableHeader)>
      </table>
      <JSXStyle
        dynamic={
          Array [
            "#fafafa",
            "auto",
          ]
        }
        id="4123435624"
      />
    </Table>
  </withClasses(Table)>
</ForwardRef(withForwardedRef(withClasses(Table)))>
`;
